lib/Optimizer/Transforms/StatePreparation.cpp

```diff
@@ -325,90 +316,81 @@ class StateDecomposer {
-      auto loc = op->getLoc();
-      builder.setInsertionPointAfter(initOp);
-      // Find the qvector alloc.
-      auto qubits = initOp.getOperand(0);
-      if (auto alloc = dyn_cast<quake::AllocaOp>(qubits.getDefiningOp())) {
-
-        // Find vector data.
-        auto data = initOp.getOperand(1);
-        auto cast = dyn_cast<cudaq::cc::CastOp>(data.getDefiningOp());
-        if (cast)
-          data = cast.getOperand();
-
-        if (auto addr =
-                dyn_cast<cudaq::cc::AddressOfOp>(data.getDefiningOp())) {
-
-          auto globalName = addr.getGlobalName();
-          auto symbol = module.lookupSymbol(globalName);
-          if (auto global = dyn_cast<cudaq::cc::GlobalOp>(symbol)) {
-            // Read state initialization data from the global array.
-            auto vec = cudaq::opt::factory::readGlobalConstantArray(global);
-
-            // Prepare state from vector data.
-            auto gateBuilder = StateGateBuilder(builder, loc, qubits);
-            auto decomposer = StateDecomposer(gateBuilder, vec, phaseThreshold);
-            decomposer.decompose();
-
-            initOp.replaceAllUsesWith(qubits);
-            toErase.push_back(initOp);
-            if (cast)
-              toErase.push_back(cast);
-            toErase.push_back(addr);
-            toErase.push_back(global);
-            return;
-          }
+class StatePrepPattern : public OpRewritePattern<quake::InitializeStateOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  explicit StatePrepPattern(MLIRContext *ctx, double phaseThreshold)
+      : OpRewritePattern(ctx), phaseThreshold(phaseThreshold) {}
+
+  LogicalResult matchAndRewrite(quake::InitializeStateOp init,
+                                PatternRewriter &rewriter) const override {
+    auto loc = init.getLoc();
+    auto qubits = init.getOperand(0);
+    if (auto alloc = dyn_cast<quake::AllocaOp>(qubits.getDefiningOp())) {
+
+      // Find vector data.
+      auto data = init.getOperand(1);
+      auto cast = data.getDefiningOp<cudaq::cc::CastOp>();
+      if (cast)
+        data = cast.getOperand();
+
+      if (auto addr = data.getDefiningOp<cudaq::cc::AddressOfOp>()) {
+        auto globalName = addr.getGlobalName();
+        auto module = init->getParentOfType<mlir::ModuleOp>();
+        auto symbol = module.lookupSymbol(globalName);
+        if (auto global = dyn_cast<cudaq::cc::GlobalOp>(symbol)) {
+
+          // Read state initialization data from the global array.
+          auto vec = cudaq::opt::factory::readGlobalConstantArray(global);
+
+          // Prepare state from vector data.
+          auto gateBuilder = StateGateBuilder(rewriter, loc, qubits);
+          auto decomposer = StateDecomposer(gateBuilder, vec, phaseThreshold);
+          decomposer.decompose();
+
+          // Use prepared qubits instead of the init state.
+          init.replaceAllUsesWith(qubits);
+          rewriter.eraseOp(init);
+          if (cast)
```

Comment:
only erasing the init op seems to be necessary (so we don't go into an infinite recursion)

Reply:
done